### ISSUE TITLE
Chained pipes in sequential order

### ISSUE THREAD

**Issue by danielfaust** [jdeferred/jdeferred#57]
Hi, 
I'm trying to chain multiple pipes.

I'm facing the issue that the second pipe, which the code in block 3, is running in parallel with the block 2.

I know that I apparently would need to pipe the code in block 3 to the promise p2, but that doesn't feel right. Wouldn't it make more sense for p2 to replace p during the DonePipe?

How can I archive what I want, without nesting block 3 into the DonePipe.pipeDone of block 2? This means, how can I access p3 from outside of the wrapper it's running in, that is, at the same level as the promise p?

Could I pre-build the promises without making them execute, and then execute them later when required?

Thanks in advance.

```java
final AndroidDeferredManager dm = new AndroidDeferredManager();

// block 1
Promise p = dm.when(new DeferredAsyncTask<Void, Integer, Object>() {
    @Override protected String doInBackgroundSafe(Void... nil) throws Exception {
        Log.v(TAG, "+++++++ BACKGROUND SLEEP");
        Thread.sleep(6000);
        Log.v(TAG, "+++++++ BACKGROUND RESUME");
        return "done";
    }
}).always(new AlwaysCallback<Object, Throwable>() {
    @Override
    public void onAlways(Promise.State state, Object o, Throwable throwable) {
        Log.v(TAG, "+++++++ ALWAYS " + state.toString() + " o: " + o + " t:" + throwable);
    }
});

// block 2
p.then(new DonePipe() {
    @Override public Promise pipeDone(Object o) {
        Promise p2 = dm.when(new DeferredAsyncTask<Void, Integer, Object>() {
            @Override protected String doInBackgroundSafe(Void... nil) throws Exception {
                Log.v(TAG, "+++++++ BACKGROUND SLEEP 2");
                Thread.sleep(4000);
                int i = 1 / 0;
                Log.v(TAG, "+++++++ BACKGROUND RESUME 2");
                return "done 2";
            }
        });
        return p2;
    }
}).always(new AlwaysCallback<Object, Throwable>() {
    @Override
    public void onAlways(Promise.State state, Object o, Throwable throwable) {
        Log.v(TAG, "+++++++ ALWAYS 2 " + state.toString() + " o: " + o + " t:" + throwable);
    }
});

// block 3
p.then(new DonePipe() {
    @Override public Promise pipeDone(Object o) {
        Promise p3 = dm.when(new DeferredAsyncTask<Void, Integer, Object>() {
            @Override protected String doInBackgroundSafe(Void... nil) throws Exception {
                Log.v(TAG, "+++++++ BACKGROUND SLEEP 3");
                Thread.sleep(4000);
                Log.v(TAG, "+++++++ BACKGROUND RESUME 3");
                return "done 3";
            }
        });
        return p3;
    }
}).always(new AlwaysCallback<Object, Throwable>() {
    @Override
    public void onAlways(Promise.State state, Object o, Throwable throwable) {
        Log.v(TAG, "+++++++ ALWAYS 3 " + state.toString() + " o: " + o + " t:" + throwable);
    }
});
```

**Comment by danielfaust** [comment]
Hi, I think I'm starting to understand the problem.

I changed the code a bit and the code below works as expected. Everything is chained and gets run sequentially. The **problem** is that when block 2 causes an **exception**, then **the piped deferred in block 3 won't get executed**. It will jump right into the .always of block 3 replicating the throwable of block 2.

Is that a standard, expected behavior? To me this feels like a bug or limitation, but maybe I'm just not understanding the consequences of not doing it this way.

```java
// === block 1
Promise p1 = dm.when(new DeferredAsyncTask<Void, Integer, Object>() {
    @Override protected String doInBackgroundSafe(Void... nil) throws Exception {
        Log.v(TAG, "+++++++ BACKGROUND SLEEP 1");
        Thread.sleep(6000);
        Log.v(TAG, "+++++++ BACKGROUND RESUME 1");
        return "done";
    }
}).always(new AlwaysCallback<Object, Throwable>() {
    @Override public void onAlways(Promise.State state, Object o, Throwable throwable) {
        Log.v(TAG, "+++++++ ALWAYS 1 " + state.toString() + " o: " + o + " t:" + throwable);
    }
}).then(new DonePipe() { // === block 2
    @Override public Promise pipeDone(Object o) {
        Log.v(TAG, "+++++++ PIPING 2");
        return dm.when(new DeferredAsyncTask<Void, Integer, Object>() {
            @Override protected String doInBackgroundSafe(Void... nil) throws Exception {
                Log.v(TAG, "+++++++ BACKGROUND SLEEP 2");
                Thread.sleep(6000);
                //int i = 1 / 0; <<<<<<=============================THIS HERE, PROBLEM
                Log.v(TAG, "+++++++ BACKGROUND RESUME 2");
                return "done 2";
            }
        });
    }
}).always(new AlwaysCallback<Object, Throwable>() {
    @Override
    public void onAlways(Promise.State state, Object o, Throwable throwable) {
        Log.v(TAG, "+++++++ ALWAYS 2 " + state.toString() + " o: " + o + " t:" + throwable);
    }
}).then(new DonePipe() { // === block 3
    @Override public Promise pipeDone(Object o) {
        Log.v(TAG, "+++++++ PIPING 3");
        return dm.when(new DeferredAsyncTask<Void, Integer, Object>() {
            @Override protected String doInBackgroundSafe(Void... nil) throws Exception {
                Log.v(TAG, "+++++++ BACKGROUND SLEEP 3");
                Thread.sleep(6000);
                Log.v(TAG, "+++++++ BACKGROUND RESUME 3");
                return "done 3";
            }
        });
    }
}).always(new AlwaysCallback<Object, Throwable>() {
    @Override
    public void onAlways(Promise.State state, Object o, Throwable throwable) {
        Log.v(TAG, "+++++++ ALWAYS 3 " + state.toString() + " o: " + o + " t:" + throwable);
    }
});
```

**Comment by danielfaust** [comment]
Would it be correct to assume this behavior is expected, because it is my job to take care that the DeferredAsyncTask in block 2 does not throw, ie by handling the exceptions in there, unless is is some kind of really bad exception where I think I need to abort everything, including any piped deferreds which are chained after the exception point? That would at least make some sense.

**Comment by danielfaust** [comment]
OK, I got it. I hadn't seen the FailPipe (found that through the jQuery docs). So if I add a FailPipe in block 3 and duplicate the DeferredAsyncTask of the DonePipe in block 3, then I can recover from an exception, that is, I can get a "ALWAYS 3 RESOLVED o: done 3 t:null" even if block 2 fails.

That's it, it's a beautiful framework. Just what I was looking for. Thanks a lot.
